python/strings_udf/strings_udf/lowering.py

```diff
@@ -107,6 +108,35 @@ def cast_string_literal_to_string_view(context, builder, fromty, toty, val):
     return sv._getvalue()
 
 
+@cuda_lowering_registry.lower_cast(string_view, udf_string)
+def cast_string_view_to_udf_string(context, builder, fromty, toty, val):
+    sv_ptr = builder.alloca(default_manager[fromty].get_value_type())
+    udf_str_ptr = builder.alloca(default_manager[toty].get_value_type())
+    builder.store(val, sv_ptr)
+    _ = context.compile_internal(
+        builder,
+        call_create_udf_string_from_string_view,
+        nb_signature(types.void, _STR_VIEW_PTR, types.CPointer(udf_string)),
+        (sv_ptr, udf_str_ptr),
+    )
+    result = cgutils.create_struct_proxy(udf_string)(
+        context, builder, value=builder.load(udf_str_ptr)
+    )
+
+    return result._getvalue()
+
+
+# utilities
+_create_udf_string_from_string_view = cuda.declare_device(
+    "udf_string_from_string_view",
+    types.void(types.CPointer(string_view), types.CPointer(udf_string)),
+)
+
+
+def call_create_udf_string_from_string_view(sv, udf_str):
+    _create_udf_string_from_string_view(sv, udf_str)
```

Comment:
It feels pretty magical that the UDFString type in _typing.py and these 30 lines of code are all that it takes to enable the return type now!

Reply:
I was a bit surprised too. Here's the magic: Assume you had a normal numba kernel that you were compiling

```python
def kernel(input, output)
    tid = cuda.threadIdx.x
    output[tid] = input[tid]
```
Now assume that we specialize this kernel with the input types (`float32*`, `int32*`). Numba knows enough to insert a cast from float to int in the IR before setting the final item into the output array. Ultimately this would appear as some kind of `cvt` instruction between primitive scalars in the PTX code.

What we're doing here is actually the same. The single test we have added is essentially:

```python
def f(string):
    return string
```

The kernel we're launching then is not much more than
```python
def kernel(input, output)
    tid = cuda.threadIdx.x
    output[tid] = f(input[tid])
```
typed as (`string_view*`, `udf_string*`). 

This means that all we really need to do is implement the cast from `string_view` to `udf_string`, which I added here as eventually translating down through the shim layer to a call to `udf_string`'s actual constructor. It ends up being a function call instruction in the PTX code (which is all that the `cuda.declare_device` functions _really_ add) and the implementation is linked in later from the main PTX blob that we ship. It's a black box to numba, all it knows is to allocate enough memory on the stack for a `udf_string` sized object, pass that pointer to the shim function and return whatever comes out.

Lots of stuff for free thanks to numba.